group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,959 @@
+            }
+
+            log.debug("Update last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.getOrCreateSnapshot(offset);
+        }
+
+        /**
+         * Reverts the last written offset. This also reverts the snapshot
+         * registry to this offset. All the changes applied after the offset
+         * are lost.
+         *
+         * @param offset The offset to revert to.
+         */
+        private void revertLastWrittenOffset(
+            long offset
+        ) {
+            log.debug("Revert last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.revertToSnapshot(offset);
+        }
+
+        /**
+         * Updates the last committed offset. This completes all the deferred
+         * events waiting on this offset. This also cleanups all the snapshots
+         * prior to this offset.
+         *
+         * @param offset The new last committed offset.
+         */
+        private void updateLastCommittedOffset(
+            long offset
+        ) {
+            if (offset <= lastCommittedOffset) {
+                throw new IllegalStateException("New committed offset " + offset + " of " + tp +
+                    " must be larger than " + lastCommittedOffset + ".");
+            }
+
+            log.debug("Update committed offset of {} to {}.", tp, offset);
+            lastCommittedOffset = offset;
+            deferredEventQueue.completeUpTo(offset);
+            snapshotRegistry.deleteSnapshotsUpTo(offset);
+        }
+
+        /**
+         * Closes this coordinator. All the deferred events are failed
+         * with a NOT_COORDINATOR error.
+         */
+        private void close() {
+            log.debug("Closing coordinator for " + tp + ".");
+            state = CoordinatorState.CLOSED;
+            deferredEventQueue.failAll(Errors.NOT_COORDINATOR.exception());
+            coordinator.onUnloaded();
+            log.debug("Coordinator for " + tp + " closed.");
+        }
+    }
+
+    /**
+     * A coordinator write operation.
+     *
+     * @param <S> The type of the coordinator state machine.
+     * @param <T> The type of the response.
+     * @param <U> The type of the records.
+     */
+    interface CoordinatorWriteOperation<S, T, U> {
+        /**
+         * Generates the records needed to implement this coordinator write operation. In general,
+         * this operation should not modify the hard state of the coordinator. That modifications
+         * will happen later on, when the records generated by this function are applied to the
+         * coordinator.
+         *
+         * @param state The coordinator state machine.
+         * @return A result containing a list of records and the RPC result.
+         * @throws KafkaException
+         */
+        CoordinatorResult<T, U> generateRecordsAndResult(S state) throws KafkaException;
```

Comment:
nit: i think it's a bit confusing to use `state` since we have CoordinatorState. can we use stateMachine?

Reply:
I think that I will use `coordinator` to stay consistent with the rest of the class.